eo-phi-normalizer/app/Main.hs

```diff
@@ -1,111 +1,222 @@
-      ruleSet <- parseRuleSetFromFile path
+  opts <- customExecParser pprefs cliOpts
+  case opts of
+    CLI'MetricsPhi' CLI'MetricsPhi{..} -> do
+      let ?parserContext = Optparse.Context metricsCommandName metricsParserInfo
+      program' <- getProgram inputFile program
+      (logStrLn, _) <- getLoggers outputFile
+      let metrics = collectMetrics program'
+      logStrLn $ encodeToJSONString metrics
+    CLI'TransformPhi' CLI'TransformPhi{..} -> do
+      let ?parserContext = Optparse.Context transformCommandName transformParserInfo
+      program' <- getProgram inputFile program
+      (logStrLn, logStr) <- getLoggers outputFile
+      ruleSet <- parseRuleSetFromFile rulesPath
       unless (single || json) $ logStrLn ruleSet.title
-      src <- maybe getContents readFile inPath
-      let progOrError = parseProgram src
-      case progOrError of
-        Left err -> error ("An error occurred parsing the input program: " <> err)
-        Right input@(Program bindings) -> do
-          let results
-                | chain = applyRulesChain (Context (convertRule <$> ruleSet.rules) [Formation bindings]) (Formation bindings)
-                | otherwise = pure <$> applyRules (Context (convertRule <$> ruleSet.rules) [Formation bindings]) (Formation bindings)
-              uniqueResults = nub results
-              totalResults = length uniqueResults
-          when (null uniqueResults || null (head uniqueResults)) $ error "Could not normalize the program"
-          let printFormationAsProgramOrObject = \case
-                Formation bindings' -> printTree $ Program bindings'
-                x -> printTree x
-          if single
+      let Program bindings = program'
+          results
+            | chain = applyRulesChain (Common.Context (convertRule <$> ruleSet.rules) [Formation bindings]) (Formation bindings)
+            | otherwise = pure <$> applyRules (Common.Context (convertRule <$> ruleSet.rules) [Formation bindings]) (Formation bindings)
+          uniqueResults = nub results
+          totalResults = length uniqueResults
+      when (null uniqueResults || null (head uniqueResults)) $ die [i|Could not normalize the #{_PROGRAM}.|]
+      let printAsProgramOrAsObject = \case
+            Formation bindings' -> printTree $ Program bindings'
+            x -> printTree x
+      if single
+        then
+          logStrLn
+            . (if json then encodeToJSONString else id)
+            . printAsProgramOrAsObject
+            $ head (head uniqueResults)
+        else do
+          if json
             then
-              logStrLn
-                . (if json then encodeString else id)
-                . printFormationAsProgramOrObject
-                $ head (head uniqueResults)
+              logStrLn . encodeToJSONString $
+                StructuredJSON
+                  { input = printTree program'
+                  , output = (printAsProgramOrAsObject <$>) <$> results
+                  }
             else do
```

Comment:
Too much nesting of `if`s leads to higher cognitive load when reading the code. Consider refactoring somehow

Reply:
Didn't think 1 nested if is too much.